src/field/FieldIntegrator.i.hh

```diff
@@ -0,0 +1,265 @@
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file FieldIntegrator.i.hh
+//---------------------------------------------------------------------------//
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with shared data and the stepper.
+ */
+CELER_FUNCTION
+FieldIntegrator::FieldIntegrator(const FieldParamsPointers& shared,
+                                 RungeKutta&                stepper)
+    : shared_(shared), stepper_(stepper)
+{
+    CELER_ENSURE(shared_);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Adaptive step control based on G4MagIntegratorDriver
+ */
+CELER_FUNCTION
+real_type FieldIntegrator::advance_chord_limited(real_type hstep, ode_type& y)
+{
+    ode_type  yend         = y;
+    real_type curve_length = 0;
+
+    real_type dyerr;
+    real_type h = this->find_next_chord(hstep, y, yend, dyerr);
+
+    // evaluate a relative error
+    real_type ratio = dyerr / (shared_.epsilon_step * h);
+
+    if (ratio < 1.0)
+    {
+        // Accept this accuracy and update the state
+        y = yend;
+    }
+    else
+    {
+        // Compute the hnext for accuracy_advance
+        real_type hnext = this->new_step_size(h, ratio);
+
+        // Advance more accurately to the "end of chord"
+        bool is_good = this->accurate_advance(h, y, curve_length, hnext);
+        if (!is_good)
+        {
+            h = curve_length;
+        }
+    }
+    return h;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * find the next chord and return a step length taken and updates the state
+ */
+CELER_FUNCTION
+real_type FieldIntegrator::find_next_chord(real_type       hstep,
+                                           const ode_type& y,
+                                           ode_type&       yend,
+                                           real_type&      dyerr)
+{
+    // Try with the proposed step
+    real_type h = hstep;
+
+    ode_type dydx;
+    ode_rhs(y, dydx);
+
+    bool converged = false;
+    unsigned int step_countdown = shared_.max_nsteps;
+    do {
+        // Always start from the initial point
+        yend = y;
+        real_type dchord_step;
+
+        dyerr = this->quick_advance(h, yend, dydx, dchord_step);
+
+        // Exit if the distance to the chord is small than the reference
+        converged = (dchord_step <= shared_.delta_chord); 
+ 
+       if(!converged)  {
+      	    if (CELER_UNLIKELY(--step_countdown == 0))
+	    {
+	        // Failed to converge, handle rare cases here
+	        break;
+	    }
+            // try a reduced step size, but not more than a factor 2
+            h *= std::fmax(sqrt(shared_.delta_chord / dchord_step), 0.5);
+	}
+    } while (!converged);
```

Comment:
Originally your loops were written as though the point was to count a certain
number of steps, and converging early was an afterthought.
```c++
setup();
for (int i = 0; i < max_steps; ++i)
{
    step();
    if (is_converged())
        break;
    pre_step();
}
```

After my half-baked suggestion, the loops are now written as though the point
is to converge, but there's extra complication from having to check for
convergence twice on every loop, and also if you happen to converge at the same
point the step limiter is hit, you're testing the convergence for nothing.
```c++
bool converged;
int remaining_steps = max_steps;
setup();
do
{
    step();
    converged = is_converged();
    if (!converged)
    {
        if (--remaining_steps == 0)
            break;
        pre_step();
    }
} while (!converged);
```

I think you should flip this around by refactoring the loop so that the
"before-next-step" always happens inside the step, before the next step. The
first iteration happens outside the loop, and then the remaining iterations are
inside the loop.
```c++
setup();
step();
int remaining_steps = max_steps;
while (!is_converged() && --remaining_steps > 0)
{
    pre_step();
    step();
}
```

The first version has two jumps (`break` and `for`) and two conditonals (one as
part of the `for`); the second has two jumps and three conditionals; the last
has a single conditional and a single jump.

Minimizing jumps and conditionals has long been recognized as important to
reducing code complexity and unexpected bugs, so this last version is not only
the clearest in intent but also the least likely to be buggy.

Reply:
Okay, the last is a good suggestion and codes are changed accordingly - I also considerred something similar logic in the beginning, but worried about the code redundancy. Isn't calling step(...) twice adding number of instructions and memory copies (i.e., passing arguments)?

Reply:
That's a good point, I wasn't considering the optimization angle; I was mainly thinking about the readability. Compilers are pretty smart about optimizing loops if we give them enough information.... I did create a CPU version of these three versions (plus a fourth that uses a flag to switch from setup to pre-step after the first loop): https://github.com/sethrj/testsnippets/tree/master/iteration-to-convergence . It's not going to be directly applicable to the GPU version, but the `do/while` version (the second one) has the fewest instructions even though it's the hardest to read. I guess the ideal readability plus performance would be to have a well-tested `template<class T> bool solve(T&& solver)` and define an  interface for `T` that has `step()`, `update_step()`, `converged()`, etc.... then just call `is_converged = solve(stepper);` in each of these places. The generic `solve` there could be as ugly as it needed for efficiency because it's only defined one place and we  test all the edge cases.

The mag field stuff as it is has enough small helper classes already, though, so I'd say that this iteration should focus on readability... for our next performance testing round let's try improving the loop structure.